engine/table/src/main/java/io/deephaven/engine/table/impl/perf/QueryPerformanceNugget.java

```diff
@@ -4,121 +4,234 @@
+     * @param operationNumber A unique identifier for the operation within a query evaluation
+     * @param parentOperationNumber The unique identifier of the parent operation or {@link QueryConstants#NULL_INT} if
+     *        none
      * @param depth Depth in the evaluation chain for the respective operation
      * @param description The operation description
      * @param isUser Whether this is a "user" nugget or one created by the system
      * @param inputSize The size of the input data
+     * @param onCloseCallback A callback that is invoked when the nugget is closed. It returns whether the nugget should
+     *        be logged.
      */
-    QueryPerformanceNugget(final int evaluationNumber, final int depth,
-            final String description, final boolean isUser, final long inputSize) {
+    protected QueryPerformanceNugget(
+            final long evaluationNumber,
+            final long parentEvaluationNumber,
+            final int operationNumber,
+            final int parentOperationNumber,
+            final int depth,
+            @NotNull final String description,
+            @Nullable final String sessionId,
+            final boolean isUser,
+            final long inputSize,
+            @NotNull final Predicate<QueryPerformanceNugget> onCloseCallback) {
         startMemorySample = new RuntimeMemory.Sample();
         endMemorySample = new RuntimeMemory.Sample();
         this.evaluationNumber = evaluationNumber;
+        this.parentEvaluationNumber = parentEvaluationNumber;
+        this.operationNumber = operationNumber;
+        this.parentOperationNumber = parentOperationNumber;
         this.depth = depth;
         if (description.length() > MAX_DESCRIPTION_LENGTH) {
             this.description = description.substring(0, MAX_DESCRIPTION_LENGTH) + " ... [truncated "
                     + (description.length() - MAX_DESCRIPTION_LENGTH) + " bytes]";
         } else {
             this.description = description;
         }
+        this.sessionId = sessionId;
         this.isUser = isUser;
         this.inputSize = inputSize;
+        this.onCloseCallback = onCloseCallback;
 
         authContext = ExecutionContext.getContext().getAuthContext();
         callerLine = QueryPerformanceRecorder.getCallerLine();
 
         final RuntimeMemory runtimeMemory = RuntimeMemory.getInstance();
         runtimeMemory.read(startMemorySample);
 
-        startAllocatedBytes = ThreadProfiler.DEFAULT.getCurrentThreadAllocatedBytes();
-        startPoolAllocatedBytes = QueryPerformanceRecorder.getPoolAllocatedBytesForCurrentThread();
-
-        startClockTime = System.currentTimeMillis();
-        startTimeNanos = System.nanoTime();
-
-        startCpuNanos = ThreadProfiler.DEFAULT.getCurrentThreadCpuTime();
-        startUserCpuNanos = ThreadProfiler.DEFAULT.getCurrentThreadUserTime();
-
-        state = QueryState.RUNNING;
-        shouldLogMeAndStackParents = false;
+        state = QueryState.NOT_STARTED;
```

Comment:
I think we should do a quick audit of instance members and constructors.
1. I don't think we should mix inline initialization with constructor initialization.
2. We should make sure the order of members matches the order of initializations in the constructors, so it's easy to verify that we didn't make any mistakes.
3. I prefer not to explicitly initialize to default values, maybe. Or, do so consistently.